### PR TITLE
ci: semver-checks for non-additive cargo features

### LINKED ISSUE #3001: Add a semver check checking no features -> all features rather than across versions
This would detect if we accidentally make a feature non-additive. Ideally we would test all combinations but that's impractical for many features.

### LINKED ISSUE #3023: ci: pin dynamically `cargo-semver-checks` version
One of the TODO lefts in the already huge #3003.

## What we need?

1. a file `.github/workflows/cargo-semver-version` similar to `nightly-version` that tracks the pinned version of `cargo-semver-checks`.
2. a weekly `cron-weekly-update-cargo-semver-checks.yml` that checks if we need to update `cargo-semver-checks` and makes a PR (similar to other `cron-*.yml` jobs)

Some ideas:

- We can try to get some information of the latest version with https://github.com/obi1kenobi/cargo-semver-checks/releases/latest
- We can also try to use the [`crates.io` API](https://crates.io/api/v1/crates) (my preferred way).

### PR DESCRIPTION
**Pull request by storopoli**
Closes #3001.
Closes #3023.

## Summary

Adds new CI Job that checks for semver breaks in `no-default-features` against `all-features`.
A fail would represent that we are inserting non-additive cargo features.

Credits to @Kixunil for the amazing idea.

This PR does:

1. fix a non-additive feature in `bitcoin-io`;
1. move the current `semver-checks` to `semver-checks-pr` since it checks if the PR is introducing classical semver-checks;
1. adds a new `semver-checks-feature`;
1. Add `-n` to `unzip` in `semver-checks`;
1. pins `rustc` stable and updates in a cron job weekly on Fri; and
1. pins `cargo-semver-checks` to `0.33.0` and updates in a cron job weekly on Sat.

## Implementation notes

We don't need nightly (and somehow it fails) but if we incorporate the ENV [`RUSTC_BOOTSTRAP=1` as hardcoded in the cargo-semver-checks codebase](https://github.com/obi1kenobi/cargo-semver-checks/blob/50b93599df92e38ddbcec0754837e8b6fd8c77ad/src/rustdoc_cmd.rs#L110C17-L110C39) and run `cargo` with the stable toolchain it works.

To get the `cargo-semver-checks` version we use the [`crates.io` API](https://crates.io/api/v1/crates/cargo-semver-checks) along with some quite intelligible `jg` magic (NOTE: `jq` is available in GH's `ubuntu-latest` image).

### CODE REVIEW

**Comment by storopoli** [comment]
Apparently, this already caught a non-additive cargo feature in `bitcoin-io`:

```
FAIL [   0.001s]       major        auto_trait_impl_removed

<......>

--- failure auto_trait_impl_removed: auto trait no longer implemented ---

Description:
A public type has stopped implementing one or more auto traits. This can break downstream code that depends on the traits being implemented.
        ref: https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.32.0/src/lints/auto_trait_impl_removed.ron

Failed in:
  type Error is no longer UnwindSafe, in io/src/error.rs:7
  type Error is no longer RefUnwindSafe, in io/src/error.rs:7
     Summary semver requires new major version: 1 major and 0 minor checks failed
    Finished [   0.022s] <unknown>
```

File: contrib/check-semver-feature.sh
Comment (obi1kenobi) [inline contrib/check-semver-feature.sh:71]: Consider changing the file name pattern to be different from the one used in the breaking change checking script, so that if the two steps end up running inside the same job runner, you don't get confusing errors.
```suggestion
    cargo semver-checks -v --baseline-rustdoc "$crate-no-default-features.json" --current-rustdoc "$crate-all-features.json" > "$crate--additive-features.txt" 2>&1
```

File: contrib/check-semver-feature.sh
Comment (obi1kenobi) [inline contrib/check-semver-feature.sh:110]: Corresponding change to the suggestion in my other comment, if you decide to follow it:
```suggestion
    for file in *additive-features.txt; do
```

File: contrib/check-semver-feature.sh
Comment (obi1kenobi) [inline contrib/check-semver-feature.sh:52]: typo fix:
```suggestion
    check_for_non_additive_cargo_features
```

File: contrib/check-semver-feature.sh
Comment (obi1kenobi) [inline contrib/check-semver-feature.sh:109]: typo fix:
```suggestion
check_for_non_additive_cargo_features() {
```

**Comment by obi1kenobi** [comment]
> We don't need nightly (and somehow it fails)

The rustdoc JSON format has frequent breaking changes: e.g. one last week, and very probably one more today or tomorrow. We don't always support the latest nightly's format because that would be a very heavy maintenance burden. I think my limited time and energy would benefit the community more if I dedicated it to other work instead.

**Comment by Kixunil** [comment]
Sounds like `cargo-semver-checks` should declare which version it supports and then we just pin that.

**Comment by storopoli** [comment]
> Sounds like `cargo-semver-checks` should declare which version it supports and then we just pin that.

stable, that's why I am running this on stable.

I think I should also update the original semver to run on stable and use the `RUSTC_BOOTSTRAP` env var as well. (another PR)

**Comment by Kixunil** [comment]
FYI to fix that error this new check found you need to add `_not_unwind_safe: core::marker::PhantomData<(&'static mut (), core::cell::UnsafeCell<()>)>,` field to the mentioned error struct. This is a breaking change for IO but it was in principle broken already.

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:41]: Shouldn't say semver, it should say "Non additive cargo features"

Comment (storopoli) [reply .github/workflows/semver-checks.yml]: Done, thanks!

**Comment by storopoli** [comment]
> FYI to fix that error this new check found you need to add `_not_unwind_safe: core::marker::PhantomData<(&'static mut (), core::cell::UnsafeCell<()>)>,` field to the mentioned error struct. This is a breaking change for IO but it was in principle broken already.

Done in 1164ed9e7b23262f391d566ceee6a8df677ab8db, since we use merge commits for the PR strategy, I took the liberty of adding to this PR.

**Comment by Kixunil** [comment]
@obi1kenobi I meant we need to sync them somehow. I suggest adding a `--print-rust-json-version` flag to print the current version so it can be synced in CI automatically.

**Comment by obi1kenobi** [comment]
> @obi1kenobi I meant we need to sync them somehow. I suggest adding a `--print-rust-json-version` flag to print the current version so it can be synced in CI automatically.

Sorry, I'm confused. Could you elaborate on the problem and how the output of such a flag would be helpful? I'm not sure what needs to be synced and isn't currently.

My understanding of this PR is that if current-stable Rust is used, everything should just work.

**Comment by Kixunil** [comment]
> Sorry, I'm confused. Could you elaborate on the problem and how the output of such a flag would be helpful? I'm not sure what needs to be synced and isn't currently.

Whenever we update one we need to update the other. The PR currently works but in a few months "latest stable" will be a different version than it is now and it may change the Json output breaking our CI. That' why we pin things. We can obviously pin both (and we should!) but we also want to update the pin periodically to not stay behind. A failed periodic update doesn't interfere with our work on merging PRs which is really helpful since we don't have to wait until a solution is found (in case it's not obvious) or a new compiler/tool version (if it's a their bug) and may avoid costly rebases of PRs.

To do the periodic update automatically int the CI we need an automated way to assign compiler version to the tool version. Having an option that just prints it is easiest. An alternative would be to publish a table at a stable URL that gets updated with each release. I think the switch is a simpler approach.

**Comment by obi1kenobi** [comment]
> Whenever we update one we need to update the other. The PR currently works but in a few months "latest stable" will be a different version than it is now and it may change the Json output breaking our CI. That' why we pin things. We can obviously pin both (and we should!) but we also want to update the pin periodically to not stay behind.

The current code in the PR installs latest stable and latest `cargo-semver-checks`, which are guaranteed to work with each other and will not break:
https://github.com/rust-bitcoin/rust-bitcoin/pull/3003/files#diff-8c3de78a6f07b22f56457fe701ccc74e767b3e22be85276c500eb1a81c01281eR48-R53

I recommend against pinning to a specific `cargo-semver-checks` version in most cases, since it usually ends up being more work and more trouble than it's worth.

Is there a specific reason you wanted to pin it?

**Comment by apoelstra** [comment]
@obi1kenobi because if you add a new lint to cargo-semver-checks, which reveals breakage in our codebase which results in a red X on CI, then all of our release branches (which do not receive regular updates, especially with API-breaking things) will permanently have a red X on them. Furthermore, every one of our current PRs will have a red X on them forcing us to drop everything to address the new cargo-semver-checks issue. (And if the issue is a bug in cargo-semver-checks, then we will wind up needing to pin it anyway while the upstream issue is resolved.)

Furthermore, if we find an old issue that CI should have caught in the past, we will have no way of re-running (or even knowing) what "CI" was back then, which makes our forensics harder.

If we just pin everything and then use cronjobs to update the pins, that fixes all of the above issues. New cargo-semver-checks issues will appear as a red X on the "update pin" PR and we can address it at our leisure.

@Kixunil honestly I don't think we need upstream support to do this. If @obi1kenobi says it always supports "the latest stable" then our cronjob can just query "the latest stable" and use `rustc -V` to get its exact version.

**Comment by obi1kenobi** [comment]
Your call on pinning, of course! :+1:

Just keep in mind that "the latest stable" policy means that the _latest_ cargo-semver-checks supports the stable that was latest at the time of its release. It does not mean that an arbitrary past version of cargo-semver-checks will always support today's latest stable (i.e. a Rust released "in the future" relative to that cargo-semver-checks version). This is probably obvious, but I wanted to err on the side of caution and mention it just in case :)

**Comment by Kixunil** [comment]
@obi1kenobi what @apoelstra said plus I want to mention that problems with unpinned stuff happened already multiple times. Stability is very helpful and since we pin other things already anyway it's not that big deal.

> @Kixunil honestly I don't think we need upstream support to do this. If @obi1kenobi says it always supports "the latest stable" then our cronjob can just query "the latest stable" and use `rustc -V` to get its exact version.

Good point, though it assumes `cargo-semver-checks` will be released sufficiently soon after rustc release. Otherwise we'll get transient failures on the update PR which would be a bit annoying. @obi1kenobi do you have any policy on this? E.g. "release the version no later than 12 hours after Rust is released" would allow us to set the cron to update at certain time (since Rust also updates at certain time) as an easy way to avoid the annoyance.

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:53]: As discussed, we need to pin the version (and the Rust version). Sorry for not realizing this sooner.

**Comment by obi1kenobi** [comment]
> Oh that's great then! So does it mean both beta and stable are supported and it can detect the correct version and use it dynamically?

In your case, you're generating the rustdoc JSON with some version of Rust, and cargo-semver-checks will detect the format version and (if supported) adjust itself automatically, yes. The current-stable and current-beta versions at the time of the cargo-semver-checks' version release are both supported, always.

If you were having cargo-semver-checks itself generate the rustdoc JSON, it would use whatever Rust version is the default as configured in rustup. It won't switch to another version since I felt that was too magical and would be the sort of thing that causes a few hours of debugging for someone sooner or later. Better to do "the obvious thing" even if it results in a (very quick and obvious-to-fix) error.

File: .github/workflows/semver-checks.yml
Comment (storopoli) [reply .github/workflows/semver-checks.yml:53]: I think we need only to pin the rust version right? As per

> In your case, you're generating the rustdoc JSON with some version of Rust, and cargo-semver-checks will detect the format version and (if supported) adjust itself automatically, yes.
>
> @obi1kenobi

Second, this is a dynamic pin, right? Same as we do with `$(cat ./nightly-version)`?

**Comment by Kixunil** [comment]
Thanks @obi1kenobi that all sounds great!

So @storopoli I think we should have a files storing latest stable version and semver checks version, then use those in the script and attempt to update them to the latest ones weekly on Thursday (not sure about the time, maybe try Friday?)

**Comment by storopoli** [comment]
Ok rust stable for semver-checks is pinned and has a cron weekly job on Fri in 5d163b1.
Do we want to pin `cargo-semver-checks` as well?

**Comment by storopoli** [comment]
This is almost done.
What is interesting is that this is the first real test for the semver detecting breaks since 1f7f080a7146138e575f552afca0408e10e85cef breaks the API.
It correctly detects but cannot add a label and a comment due to some strange github CI error:

> HttpError: Resource not accessible by integration

I can fix this by adding to the CI job:

```yaml
permissions: write-all
```

or by manually changing the token that the `action-add-label` will use to:

```yaml
with:
  github_token: ${{ SOME SECRET TOKEN }}
```

and to `actions-comment-pull-request`:

```ymal
with:
  GITHUB_TOKEN: ${{ SOME SECRET TOKEN }}
```

In `cron-daily-update-nightly.yml` we use the `secrets.APOELSTRA_CREATE_PR_TOKEN` to create a PR without touching the `permissions` (which is good practice).

@apoelstra do you want me to use that one? Or do you want to create a new one for comments and add breaking API labels?

**Comment by apoelstra** [comment]
@storopoli yeah, for now let's use the `secrets.APOELSTRA_CREATE_PR_TOKEN` token. This one is a bit annoying because it creates PRs as me (and then Github won't let me review or merge them) but the correct solution would be for us to create a whole new bot account just for this, and I don't feel like doing it. And I'm also not sure how those accounts are supposed to work ... do they have a password and somebody knows it and that person is just allowed to impersonate the bot? What if you want a higher bus number? Can you restrict the bot's permissions beyond restricting its API tokens?

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:37]: The version needs to be pulled from a file that's updated every Friday together with Rust version.

Comment (apoelstra) [reply .github/workflows/semver-checks.yml:37]: Maybe we should upgrade the flat nightly-version text file to a JSON file that contains the nightly version, stable version for cargo-semver-checks, version of cargo-semver-checks, ... and maybe also the stable/beta versions for the rest of CI (which aren't currently pinned because in theory they should never break us).

Comment (storopoli) [reply .github/workflows/semver-checks.yml:37]: Yes that's what I thought so. But @apoelstra said to mimic how we used to pin `cargo-public-api`.

What if we leave this as it is now, and move the dynamic pinning of `cargo-semver-checks` to an issue and add in a new PR?

**Comment by storopoli** [comment]
@apoelstra you might need to create a new secret for `rust-bitcoin/rust-bitcoin` with the following permissions:

-  issues: write (needed for commenting with the 🚨 message)
-  pull-requests: write (needed to add labels to the PR)

I tried hacking it in 0fdb456f36bd4b8f8a2dfba56c5d1249031d5344 with

```yaml
permissions:
  issues: write
  pull-requests: write
```

but to no avail.

File: .github/workflows/semver-checks.yml
Comment (apoelstra) [reply .github/workflows/semver-checks.yml:41]: I was mistaken. With `cargo-public-api` we didn't really have a plan to update the pin (actually I think we intended to just never update it because updates basically only ever broke us). Sorry about that.

**Comment by apoelstra** [comment]
The token has `org:write` access which should be sufficient. I can't find any docs about what "scopes" are needed for pull requests or issues. The docs don't mention PRs or issues at all. https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [reply .github/workflows/semver-checks.yml:41]: I would prefer not want to deal with json in scripts since it makes scripts less readable IMO. If you want it in a single file perhaps just use the shell `key=value` and source it.

Comment (apoelstra) [reply .github/workflows/semver-checks.yml:43]: That will make it difficult to parse using anything but shell.

**Comment by storopoli** [comment]
@apoelstra sorry to disturb but I would need you to check https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-the-default-github_token-permissions to see if the default permissions are restricted in a sense of only reading for the `GITHUB_TOKEN`.

Below is a screenshot of a repo of mine that has the default all permissive settings.
Just for you to have an idea of which config we are talking about.

image

---

On another note in the same docs link above, we have:

> Anyone with write access to a repository can modify the permissions granted to the `GITHUB_TOKEN`, adding or removing access as required, by editing the `permissions` key in the workflow file.

This is exactly what I am doing in https://github.com/rust-bitcoin/rust-bitcoin/blob/049cc2fbe2c9010b8ad487e362ac6b368d6a4854/.github/workflows/semver-checks.yml#L6-L8

but NOTE that I don't have **write access** to `rust-bitcoin/rust-bitcoin`.
So maybe GitHub is (rightly so) denying me to add the `permissions` key.

If that is indeed the case, then we might need to test it.
Maybe you, or someone who has write access could open a PR fixing the `bitcoin-io` non-additive cargo feature in 265057b49a48d070596775238e9f69da63f05a83 and also adding the `permission` to the workflow file This is exactly what I am doing in https://github.com/rust-bitcoin/rust-bitcoin/blob/049cc2fbe2c9010b8ad487e362ac6b368d6a4854/.github/workflows/semver-checks.yml#L6-L8

**Comment by apoelstra** [comment]
All those settings look good to me. I'll try making the PR.

https://github.com/rust-bitcoin/rust-bitcoin/pull/3020

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:8]: Needs to implement the suggestion from this article instead: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/
Have two jobs, first one does the check and uploads an artifact, the second one gets the artifact and writes the comment if the artifact says "break".

**Review by Kixunil — CHANGES_REQUESTED** [review]

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [inline .github/workflows/semver-checks-pr-label.yml:14]: Are you intentionally failing the previous job? I think it'll mark the CI as failed and we probably shouldn't do it.

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [inline .github/workflows/semver-checks-pr-label.yml:61]: Is it possible to avoid JS please? :)

Comment (storopoli) [reply .github/workflows/semver-checks-pr-label.yml:61]: The examples you linked are all in JS. To be honest even with this little exposure to JS I might need to swallow some potassium iodine over the weekend.

File: .github/workflows/semver-checks-pr-label.yml
Comment (storopoli) [reply .github/workflows/semver-checks-pr-label.yml:14]: Yes, since then this will run on every PR that triggers the check-semver (which is all of them)

Comment (storopoli) [reply .github/workflows/semver-checks-pr-label.yml]: Ok I can disable failing, lemme check...

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [reply .github/workflows/semver-checks-pr-label.yml:58]: Here's some code from my project that works with artifact without JS: https://github.com/debian-cryptoanarchy/cryptoanarchy-deb-repo-builder/blob/master/.github/workflows/ci.yml

I think you can simply store `break` or `pass` in a file, upload it in one job then download it in the other and use `if [ "`cat result`" = "break" ]` to check the result.

File: .github/workflows/semver-checks-pr-label.yml
Comment (storopoli) [inline .github/workflows/semver-checks-pr-label.yml:6]: I think we definitely need to switch to Gerrit.

Does anyone know why the hell this thing does not want to run?
I even changed the name from the other job from `Check semver breaks` -> `semver` to see if GH was getting bogged down by the spaces, but no luck!

File: .github/workflows/semver-checks-pr-label.yml
Comment (storopoli) [reply .github/workflows/semver-checks-pr-label.yml:59]: Yes, but then how do I create labels and add comments?
Hacking the `.github/labeler.yml` seems overkill.

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [reply .github/workflows/semver-checks-pr-label.yml:6]: > Does anyone know why the hell this thing does not want to run?

I'm certain it's because we haven't merged it yet - we would have the same vulnerability otherwise.

I suggest you try to develop the action in your own repo first and then make a PR once it workss.

Comment (yancyribbens) [reply .github/workflows/semver-checks-pr-label.yml]: > I will start modifying https://github.com/rust-bitcoin/rust-bitcoin/pull/3003 accordingly.

+1 - I was thinking of trying out semvar also but in a much simpler repo (first).

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [reply .github/workflows/semver-checks-pr-label.yml:58]: Use labels in your own repo?

File: .github/workflows/semver-checks-pr-label.yml
Comment (storopoli) [reply .github/workflows/semver-checks-pr-label.yml]: Success it works! I will upstream the patch to this PR.

Action: https://github.com/storopoli/rust-bitcoin/actions/runs/9913075380/job/27389270899

PR labelled: https://github.com/storopoli/rust-bitcoin/pull/9

Comment (Kixunil) [reply .github/workflows/semver-checks-pr-label.yml]: Cool! Made me realize we don't need a label for non-additive cargo features - they are a bug and thus always wrong.

**Review by Kixunil — COMMENTED** [review]
Oh, I got confused. We probably don't need the other job at all - just let the semver one fail. But we do need to fix `master`.

**Comment by storopoli** [comment]
> Oh, I got confused. We probably don't need the other job at all - just let the semver one fail. But we do need to fix `master`.

@Kixunil this is ready now. I've update the original PR description. Please read that and see if it is clear. Otherwise let me know.

**Comment by Kixunil** [comment]
I'll have to review it tomorrow. I'm super tired and unable to focus properly anymore.

**Comment by storopoli** [comment]
> I'll have to review it tomorrow. I'm super tired and unable to focus properly anymore.

No rush. Let's try to not break master again 😂

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:47]: This looks like the example from the article copied verbatim. It should be adapted to our needs, specifically, we should upload the status, not the PR number. We need to get the PR number from the even payload, otherwise griefers could make this post a comment on *all PRs* which would be highly annoying.

File: .github/workflows/semver-checks-pr-label.yml
Comment (Kixunil) [inline .github/workflows/semver-checks-pr-label.yml:37]: I think this entire step is unnecessary and vulnerable. The PR number should be in `workflow_run.pull_requests` variable though it's an array for some reason so we probably need to take the first item or iterate through it.

**Comment by Kixunil** [comment]
I've just realized why I was confused about this PR. It also fixes master and it can be hard to see which commit does so. It might be better to split fixing master into a separate PR for quicker review.

**Comment by storopoli** [comment]
> I've just realized why I was confused about this PR. It also fixes master and it can be hard to see which commit does so. It might be better to split fixing master into a separate PR for quicker review.

Yes will do that. Then we can continue improving and discussing the non-additive cargo features. Good call!

**Comment by storopoli** [comment]
Draft until we deal with #3028 and #3018

**Comment by storopoli** [comment]
Ok this is ready for another round!

**Comment by yancyribbens** [comment]
I'm kinda out of the loop on this but I might read it more later.

> API BREAKING CHANGE DETECTED

I much prefer the alerts vs the CI breakage and the diff being stuffed into git.  Although I wish there was a way to silence it after the first time or something because it gets pretty noisy after every commit.

**Comment by Kixunil** [comment]
FTR this specific PR shouldn't change any API so something wrong is going on.

We should have some marker to say "intended break" (probably a comment) which would silence the notification on pushes but still perform the check to allow us see its output. (Is it really the break we're expecting or a different one?)

**Comment by storopoli** [comment]
> FTR this specific PR shouldn't change any API so something wrong is going on.
> 
> We should have some marker to say "intended break" (probably a comment) which would silence the notification on pushes but still perform the check to allow us see its output. (Is it really the break we're expecting or a different one?)

Actually we are changing it in 7dbd44bacd42834e488093154e4e505a0ab53ce3

https://github.com/rust-bitcoin/rust-bitcoin/blob/7dbd44bacd42834e488093154e4e505a0ab53ce3/io/src/error.rs#L5-L18

> I much prefer the alerts vs the CI breakage and the diff being stuffed into git. Although I wish there was a way to silence it after the first time or something because it gets pretty noisy after every commit.

Good idea! I'll try to hack something with `actions/github-script` to suppress the warning if we already commented on the PR.

**Comment by Kixunil** [comment]
LOL, I completely forgot, sorry about the noise.

File: contrib/check-semver-feature.sh
Comment (Kixunil) [inline contrib/check-semver-feature.sh:121]: We need to fail the CI here. (`exit 1`)

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:21]: This entire job is not needed since we're using `cat` directly.

File: contrib/check-semver-feature.sh
Comment (storopoli) [reply contrib/check-semver-feature.sh:121]: Isn't this covered in the `else` here? https://github.com/rust-bitcoin/rust-bitcoin/blob/28c450febe5cbaba4defdc5421ff11138488cf1a/contrib/check-semver-feature.sh#L120-L124

File: .github/workflows/semver-checks.yml
Comment (storopoli) [reply .github/workflows/semver-checks.yml:17]: Good catch! Thanks

File: contrib/check-semver-feature.sh
Comment (Kixunil) [reply contrib/check-semver-feature.sh:121]: I got confused by the double negative :D

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [reply .github/workflows/semver-checks.yml:17]: Why is it still here?

Comment (storopoli) [reply .github/workflows/semver-checks.yml:17]: I've removed the `rust_stable_version`, that's why I told you "Good catch".
But this is needed for https://github.com/rust-bitcoin/rust-bitcoin/blob/3a19dece44c82f98a17d23bab5f373130bea6a30/.github/workflows/semver-checks.yml#L38-L39 and  https://github.com/rust-bitcoin/rust-bitcoin/blob/3a19dece44c82f98a17d23bab5f373130bea6a30/.github/workflows/semver-checks.yml#L78-L79

Comment (Kixunil) [reply .github/workflows/semver-checks.yml]: Could you please clean up the history to not add it in 2e0f15a47e1a62d4405d82fb717f8e80fa15a4e8 so it won't need to be removed later?

Comment (storopoli) [reply .github/workflows/semver-checks.yml]: Done, removed from 4edd504cb60397cfe143607ef5db8784598cbd06

**Review by Kixunil — APPROVED** [review]
ACK f76f68217b3f29cdfe3ed012091e072d64bfd95a

**Review by apoelstra — APPROVED** [review]
ACK f76f68217b3f29cdfe3ed012091e072d64bfd95a At some point we should pull the stable-version to the root with the nightly-version and use it everywhere; but can be in a separate PR

**Pull request merged**
